README.md
# exaLearnEpi
modeling of corvid-19 spread


corvid_march/HISTORY
Corvid version history

Version 0.3 - March 12, 2020

Corvid is based on FluTE (https://github.com/dlchao/FluTE).

Several parameters were adjusted to adapt the FluTE flu transmission 
to Corvid coronavirus transmission, such as incubation period,
symptomatic fraction, and viral load trajectories.
The parallel computing version was removed, so populations of more than
10,000,000 will probably not run on a laptop.

Version 0.1 or 0.2 added a "work from home" policy.
Version 0.3 is the first version the outside world has seen.
Version 0.4 fixes bug in "liberal leave" that made it happen before the policy was supposed to start and a huge bug in incubation time.
Version 0.5 adds option for home isolation after ascertainment, community contact reduction, and work policies of limited duration

= = = = = = = = = = = = = = = = 

Those are major chnages to the original code from
https://github.com/dlchao/corvid

2020-03-27, by Jan Balewski:

1) I modified epimodel.cpp in ~40 places by adding '[' and ']' for all lists so the summary file is now a regular Yaml. (I may have missed fixes for some not currently used outputs)
From now on use 'yaml' extension in the config file:
summaryfilename Summary-seattle26.yaml
because summary  is now in yaml file format.
2) I have added a print-class called in the main:
  EpiPara2yaml sum( parms.getSummaryFileName());
  sum.save(configname);
which captures the summary file name from the config-file, replaces '.ymal' --> '.meta.yaml',
and saves the config file in it in the yaml format as well.
This 2 changes allow now to read e.g. into python a well-formatted:
Summary-seattle26.yaml
Summary-seattle26.meta.yaml
where user controls this name by setting  summary filename  in the config.

3) As a bonus I added
dumpSum.py  which reads both yaml files and dumps them on the screen - just to see all works as intended.



- - - - - - - - - - - 
2020-04-03, by Jan Balewski:
 Now we are saving HD5 with ML-style 'data:tract3D' & 'labels:parBio', the accompanied titles of the columns of HD5 arrays are saved in .meta.yaml. I did not normalized parBio, no parUnit.
This 2 classes  manage it all: EpiData2hd5 and EpiParam2yaml.

Changes:
*) HD5 on Cori needs: module load cray-hdf5
*) the name of config file (after removing 'config-') is not the core-name for my 2 outputs: .meta.yaml and .h5

*) my test config: config-seattle26c includes 8 variables which have beens set be Peter H., those are added as labels to HD5 . To modify the labels saved in HD5  edit:  EpiModel::prepH5()
*) all labels are stored in HD5 as float32 

*) this method: EpiData2hd5::doDiffTract()
compute differences for the  integrals saved in Log-file. It is executed at the very end in  EpiModel::saveH5(). Disable it if you want to compare 1:1

*) to dump meta.yaml execute:   ./dumpMeta.py -N seattle26c  ( use core name)
You should see:
'hd5meta': {'fileName': 'seattle26c.h5',
             'parBio': ['TriggerDay',
                        'SchoolClosurePolicy',
                        'SchoolClosureDays',
                        'LiberalLeaveCompliance',
                        'QuarantineCompliance',
                        'WorkFromHomeCompliance',
                        'VoluntaryIsolationCompliance',
                        'AscertainedIsolationCompliance'],
             'tract3D': {'axis_name': ['census_tract', 'age_group', 'simu_day'],
                         'axis_shape': [124, 5, 40]}}}
Note, you need to pip install --user ruamel.yaml in py3 to make this script work 

*) to inspect HD5 execute:  ./dumpHD5.py -N seattle26c
You should see:
read data from hdf5: ./seattle26c.h5
read  parBio (8,) float32
read  tract3D (124, 5, 40) int32
check Labels:
[10.    1.   80.    0.81  0.82  0.83  0.84  0.85]

*) empty lines in config are now ignored
*) file HISTORY contains  this text
*) technical: the below piece of code in corvid.cpp does all the magic. It is very entangled - I admit - but I tried to keep 'my code' separately from all Dennis files - so future upgrade will be easier (and I have a lesser chance to mess-up the main code).
 - - -
  EpiModel model(parms);
  string coreName=configname;
  coreName.replace(0,7,"");
  
  EpiPara2yaml meta(coreName );
  meta.save_config(configname);
  model.prepH5(parms);
  model.run(); // execution of the  simulation
  model.saveH5(coreName);
  meta.save_hd5meta();
- - - 

Finally, this is the input I have used:
cat config-seattle26c
# basic configuration file for metro seattle
label example-seattle-26b
datafile seattle
R0 2.6
runlength 40

summaryfilename Summary-seattle26b.yaml
#logfilename Log-seattle26b.txt
#tractfilename Tracts-seattle26b.txt
#individualfilename Individuals-seattle26b.txt
#individualfile 1
randomnumberseed 1
seedinfected 5
seedinfecteddaily 0
responsethreshold 1
# from Peter H.
responseday 10
schoolclosurepolicy all
schoolclosuredays 80
liberalleave 0.81
quarantine 0.82
workfromhome 0.83
voluntaryisolation 0.84
ascertainedisolation 0.85


- - - - - - - -
2020-04-05, by Jan Balewski:

I added a EpiModel method which accumulates every day the above value in static 3D array with dims: [census_tract][age_group][simu_day]. Next, at the end of simulation I compute the difference  vs. simu_day for each fixed [census_tract][age_group]. The resulting 3D array is being saved as HD5:
symptomatic3D[census_tract][age_group][simu_day] 
where age_group_enum: [0-4, 5-18, 19-29, 30-64, 65-plus]

Detailed changes:
*)change to use commvec[i].nEverSymptomatic[j] to store final HD5: symptomatic3D[census_tract][age_group][simu_day]
*) added names of 5 age_groups to heta.yaml:
*) remove "SchoolClosurePolicy" from parBio because it is categorical and interpolation makes no sense
*) disable printouts of : 'Closing schools' etc for  track.id>2
*) Makefile: removed switch  -march=native to make CC happy on Cori
*) minute cleanup

- - - - - - - - - - - 
2020-04-07, by Jan Balewski:

*) added R0 to bioPar
*) added uniPar to Hd5 which is [-1,1]-normalized version of bioPar

sys_uncertainty/utils.py
import numpy as np
import glob
import ruamel.yaml as yaml
import warnings
warnings.simplefilter('ignore', yaml.error.MantissaNoDotYAML1_1Warning)
import h5py
import time
import sys
import matplotlib.pyplot as plt
import subprocess


def read_yaml_silent(ymlFn):
        """Non-verbose version of exalearnEpi.corvid_march.code.Util_Func.read_yaml"""
        ymlFd = open(ymlFn, 'r')
        bulk=yaml.load( ymlFd, Loader=yaml.CLoader)
        ymlFd.close()
        return bulk



class DatasetBuilder():

    def __init__(self, dataDir, dataTag, Narr, Nperjob):
        self.dataDir = dataDir # location of data files (outputs of batchRun job array)
        self.dataTag = dataTag # name tag of all batchRun outputs
        self.Narr = Narr # Number of jobs in job array
        self.Nperjob = Nperjob # number of runs per job (equals srun --ntasks in batch script)
        self.total = Narr*Nperjob
        print('Expecting  %d total runs in %s'%(self.total, self.dataDir))


    def copy_HD5s_to_cfs(self, cDir):
        print('Copying HDF5s to %s'%cDir)
        run=0
        parBios = []
        uniBios = []
        symptomatic3Ds = []

        start=time.time()
        for arrIdx in range(self.Narr):
            jobDir=self.dataDir+self.dataTag+'-'+str(arrIdx)+'/'
            print('Parsing job directory %s'%jobDir)
            for procIdx in range(self.Nperjob):
                metaFile=glob.glob(jobDir+'out_'+str(procIdx)+'/*meta.yaml')[0]
                meta = read_yaml_silent(metaFile)
                metah5 = meta['hd5meta']
                paramnames = metah5['parBio']['axis_bins']
                arraynames = metah5['symptomatic3D']['axis_name']
                age_bins = metah5['symptomatic3D']['age_group_bins']

                h5File = glob.glob(jobDir+'out_'+str(procIdx)+'/*.h5')[0]
                with h5py.File(h5File, 'r') as f:
                    parBios.append(f['parBio'][...])
                    uniBios.append(f['uniBio'][...])
                    symptomatic3Ds.append(f['symptomatic3D'][...])
                run += 1
        parBios = np.stack(parBios, axis=0)
        uniBios = np.stack(uniBios, axis=0)
        symptomatic3Ds = np.stack(symptomatic3Ds, axis=0).astype(np.ushort)
        with h5py.File(cDir+'data.h5', 'w') as f:
            f.create_dataset('parBio', data=parBios)
            print('Wrote field %s, shape %s'%('parBio', str(parBios.shape)))
            f.create_dataset('uniBio', data=uniBios)
            print('Wrote field %s, shape %s'%('uniBio', str(uniBios.shape)))
            f.create_dataset('symptomatic3D', data=symptomatic3Ds)
            print('Wrote field %s, shape %s'%('symptomatic3D', str(symptomatic3Ds.shape)))
        print('DONE')


    def copy_summaries_to_cfs(self, cDir):
        print('Copying summaries to %s'%cDir)
        parser = RunParser(self.dataDir+self.dataTag, self.Narr, self.Nperjob)
        parser.parse_runs()
        parser.save_h5(cDir+'summary.h5')

    def inspect(self):
        run=0
        parBios = []
        uniBios = []
        symptomatic3Ds = []
        
        start=time.time()
        for arrIdx in range(1):
            jobDir=self.dataDir+self.dataTag+'-'+str(arrIdx)+'/'
            print('Parsing job directory %s'%jobDir)
            for procIdx in range(self.Nperjob):
                metaFile=glob.glob(jobDir+'out_'+str(procIdx)+'/*meta.yaml')[0]
                meta = read_yaml_silent(metaFile)
                metah5 = meta['hd5meta']
                paramnames = metah5['parBio']['axis_bins']
                arraynames = metah5['symptomatic3D']['axis_name']
                age_bins = metah5['symptomatic3D']['age_group_bins']

                h5File = glob.glob(jobDir+'out_'+str(procIdx)+'/*.h5')[0]
                with h5py.File(h5File, 'r') as f:
                    parBios.append(f['parBio'][...])
                    uniBios.append(f['uniBio'][...]) 
                    symptomatic3Ds.append(f['symptomatic3D'][...])
                run += 1
        parBios = np.stack(parBios, axis=0)
        uniBios = np.stack(uniBios, axis=0)
        symptomatic3Ds = np.stack(symptomatic3Ds, axis=0)
        self.plot_params(parBios, uniBios, paramnames)
        self.plot_arrays(symptomatic3Ds, arraynames)
        print('Parsing %d runs took %f s'%(run, time.time() - start)) 


    def plot_arrays(self, arr, arrnames):
        hbins = np.arange(-0.5,134.5,1)
        centers = np.arange(134)

        hist, _ = np.histogram(arr, bins=hbins)

        plt.figure()
        plt.plot(centers, hist, 'o')
        plt.yscale('log')
        plt.xlabel('Num. new infected per age/day/tract')
        plt.ylabel('Counts')
        plt.title('symptomatic3D')
        plt.show()


    def plot_params(self, par, uni, names, nbins=20):
        numpars = par.shape[1]
        numruns = par.shape[0]
        plt.figure(figsize=(10,10))
        for idx in range(16):
           plt.subplot(4,4,idx+1)
           if idx%2 == 0:
                plt.hist(par[:,idx//2], bins=nbins)
                plt.title(names[idx//2])
           else:
                plt.hist(uni[:,idx//2], bins=nbins)
                plt.title(names[idx//2]+' (normed)')
        plt.tight_layout()
        plt.show()




class RunParser():
    """Class to parse yaml summary outputs of a set of runs into h5"""

    def __init__(self, dataDir=None, Narr=None, Nperjob=None, load_from_h5=False):
        # Usage check
        if not load_from_h5 and dataDir is None:
            print("Error: must specify either dataDir to parse a set of runs or load_from_h5 to load from an HDF5 file")
            sys.exit()
        if dataDir is not None and (Narr is None or Nperjob is None):
            print("Error: must specify Narr and Nperjob if parsing from dataDir '%s'"%dataDir)
            sys.exit()

        # fields to track
        self.fields = ['Number symptomatic (daily)',
                       'Cumulative symptomatic (daily)',
                       'Total infection attack rates by age', 
                       'Total infection attack rate', 
                       'Total infected individuals by age', 
                       'Total symptomatic attack rates by age', 
                       'Total symptomatic attack rate', 
                       'Total symptomatic individuals by age', 
                       'Total symptomatic unvaccinated individuals by age', 
                       'Total unvaccinated individuals by age', 
                       'Total individuals by age'] 

        # fields already listed by corvid as daily output
        self.fields_daily = ['Number symptomatic (daily)',
                             'Cumulative symptomatic (daily)'] 

        # Age ranges (first bin empty for mplt xtick labels)
        self.age_bins = ['', '0-4', '5-18', '19-29', '30-64', '65+'] 

        if load_from_h5:
            self.load_h5(load_from_h5)
        else: 
            self.dataDir = dataDir # location of data files (outputs of batchRun job array)
            self.Narr = Narr # Number of jobs in job array
            self.Nperjob = Nperjob # number of runs per job (equals srun --ntasks in batch script)
            self.total = Narr*Nperjob
            print('Expecting  %d total runs in %s'%(self.total, self.dataDir))
            
            metaFile = glob.glob(dataDir+'-0/out_0/*meta.yaml')[0]
            meta = read_yaml_silent(metaFile)['config']
            self.summfname = meta['summaryfilename']
            self.Ndays = meta['runlength']
            print('Summary names=%s,'%self.summfname, 'Ndays=%d'%self.Ndays)

            # Allocate arrays in the data aggregation dict
            self.allocate_arrays()
            print('Allocated arrays:')
            for k,v in self.agg.items():
                print(k, v.shape)


    def allocate_arrays(self):
        # Allocate arrays
        yamfile = self.dataDir+'-0/out_0/'+self.summfname
        dat = read_yaml_silent(yamfile)
        self.agg = {}
        for k,v in dat['DailySummary'][-1].items():
            if k in self.fields:
                if type(v) != list:
                    # Scalar field
                    self.agg[k] = np.zeros(shape=(self.total,self.Ndays,1), dtype=np.float)
                elif len(v) == self.Ndays:
                    # Fields automatically gathered as daily counts
                    self.agg[k] = np.zeros(shape=(self.total,self.Ndays,1), dtype=np.float)
                else:
                    # Others (age binned data)
                    axdim = np.array(v).shape[0]
                    self.agg[k] = np.zeros(shape=(self.total,self.Ndays,axdim), dtype=np.float)



    def parse_runs(self):
        run=0
        start=time.time()
        for arrIdx in range(self.Narr):
            jobDir=self.dataDir+'-'+str(arrIdx)+'/'
            print('Parsing job directory %s'%jobDir)
            for procIdx in range(self.Nperjob):
                yamFl=jobDir+'out_'+str(procIdx)+'/'+self.summfname
                
                days = read_yaml_silent(yamFl)['DailySummary']
                for idx, day in enumerate(days):
                    for k,v in day.items():
                        if k in self.fields:
                            if k in self.fields_daily:
                                # Don't read pre-formatted daily data until last day
                                if len(v)==self.Ndays:
                                    self.agg[k][run,:,0] = np.array(v)
                            else:
                                self.agg[k][run,idx,:] = np.array(v)
                run += 1
        print('Parsing %d runs took %f s'%(run, time.time() - start))


    def save_h5(self, fname):
            """Saves datadict fields into h5 file"""
            with h5py.File(fname, 'w') as f:
                for k,v in self.agg.items():
                    f.create_dataset(k, data=v)
            print('Saved h5 to %s'%fname)

    def load_h5(self, fname):
            """Loads h5 fields into datadict"""
            self.agg = {}
            with h5py.File(fname, 'r') as f:
                fields = list(f.keys())
                for field in fields:
                    self.agg[field] = f[field][...]
            print('From %s loaded following fields:'%fname)
            for k,v in self.agg.items():
                print(k, v.shape)

